### PR TITLE
feat: Adding filtering hashing functions

### PR DESCRIPTION
**Pull request by markvdouw**
## Instructions
 1. PR target branch should be against `development`
 2. PR title name should follow this format: https://github.com/mParticle/mparticle-workflows/blob/main/.github/workflows/pr-title-check.yml
 3. PR branch prefix should follow this format: https://github.com/mParticle/mparticle-workflows/blob/main/.github/workflows/pr-branch-check-name.yml

 ## Summary
 - Moving filtering hashing functions to utility dedicated class and replacing old ones in the code.
 - Changing tests
 - All tests must pass since the logic wasn't change, just ones in a different class function

 ## Testing Plan
 - [x] Was this tested locally? If not, explain why.
 - {explain how this has been tested, and what, if any, additional testing should be done}

 ## Reference Issue (For mParticle employees only.  Ignore if you are an outside contributor)
 - Closes https://mparticle-eng.atlassian.net/browse/SQDSDKS-5346

### CODE REVIEW

File: android-core/src/main/java/com/mparticle/internal/HashingUtility.kt
Comment (BrandonStalnaker) [inline android-core/src/main/java/com/mparticle/internal/HashingUtility.kt:22]: Why split these up into two (hashFilterAvKey & hashFilterAvValue) separate methods? Seems like an unnecessary complexity.

File: android-kit-base/src/main/java/com/mparticle/kits/KitConfiguration.java
Comment (BrandonStalnaker) [inline android-kit-base/src/main/java/com/mparticle/kits/KitConfiguration.java:613]: I'm not sure if I seen this pattern before in android. Would it be possible to remove INSTANCE or at least change it to something that matches the capitalization style of the rest of the code.

**Review by BrandonStalnaker — COMMENTED** [review]
Seems good in general. Just have a few suggestions/questions

File: android-kit-base/src/main/java/com/mparticle/kits/KitConfiguration.java
Comment (markvdouw) [reply android-kit-base/src/main/java/com/mparticle/kits/KitConfiguration.java:613]: HashingUtility is an object class in Kt (similar to a class level companion object that correspond to static).
Because functions are not called using an instance the INSTANCE wouldn't be needed if we were calling if from a Kt class also, but because KitConfiguration is a java class that is mandatory and also automatic.

File: android-core/src/main/java/com/mparticle/internal/HashingUtility.kt
Comment (markvdouw) [reply android-core/src/main/java/com/mparticle/internal/HashingUtility.kt:22]: Ben asked me to create functions in the utility class based on semantic so because conceptually there could be different ways to hash those (since conceptually those are different things) we wanted to have the separately event though the implementation is the same.

**Review by BrandonStalnaker — APPROVED** [review]
LGTM but I'd prefer if we also got @einsteinx2 approval too since he knows the implementation you're following best

File: android-core/src/main/java/com/mparticle/internal/HashingUtility.kt
Comment (einsteinx2) [reply android-core/src/main/java/com/mparticle/internal/HashingUtility.kt:22]: Yes that's correct, it's some slight code duplication but makes the purpose clearer at the call site. Plus if we ever change the hash method in the future (unlikely but possible) it's more flexible, but the main reason is the former.

File: android-kit-base/src/test/kotlin/com/mparticle/kits/KitConfigurationTest.kt
Comment (einsteinx2) [inline android-kit-base/src/test/kotlin/com/mparticle/kits/KitConfigurationTest.kt:453]:
```suggestion
        rule1.put("h", HashingUtility.hashGDPRContentPurposeKey("foo purpose"))
```
Shouldn't the `"1" +` be removed here since the new method adds it?

File: android-kit-base/src/test/kotlin/com/mparticle/kits/KitConfigurationTest.kt
Comment (einsteinx2) [inline android-kit-base/src/test/kotlin/com/mparticle/kits/KitConfigurationTest.kt:588]: For example here the `"1" +` was removed...

In that case, why did the previous error not cause a test failure? Are we not actually testing that data? If not, why include it in the test?

**Review by einsteinx2 — CHANGES_REQUESTED** [review]

**Review by einsteinx2 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
